libsnark/polynomial_commitments/kzg10_batched.hpp

```diff
@@ -0,0 +1,113 @@
+#define __LIBSNARK_POLYNOMIAL_COMMITMENTS_KZG10_BATCHED_HPP__
+
+#include "libsnark/polynomial_commitments/kzg10.hpp"
+
+/// Reference:
+/// - [GWC19]:
+///   Title: "PLONK: Permutations over Lagrange-bases for Oecumenical
+///           Noninteractive arguments of Knowledge"
+///   eprint: https://eprint.iacr.org/2019/953.pdf
+
+namespace libsnark
+{
+
+/// Batched version of KZG10, as described in Section 3 of [GWC19]. Supports
+/// evaluation of arbitrary number of polynomials at 2 points, requiring a
+/// fixed-size witness (of 2 elements of G1), which can be verified with 2
+/// pairings (similarly to KZG10) and a number of group operations proportional
+/// to the number of polynomials.
+///
+/// Note, this could be extended to the case of N evaluation points (and
+/// thereby N sets of polynomials).
+template<typename ppT> class kzg10_batched_2_point
+{
+public:
+    using Field = libff::Fr<ppT>;
+
+    /// The srs generated by the setup step. This is denoted PK in [KZG10]
+    /// section 3.2.
+    using srs = typename kzg10<ppT>::srs;
+
+    using commitment = typename kzg10<ppT>::commitment;
+
+    /// Claimed evaluations of 2 groups of polynomials, fs and gs, at 2 points
+    /// z_1 and z_2.  f_s are evaluated at z_1 and g_s are evaluated at z_2.
+    class evaluations
+    {
+    public:
+        std::vector<Field> s_1s;
+        std::vector<Field> s_2s;
+
+        evaluations(std::vector<Field> &&fs, std::vector<Field> &&gs);
+    };
+
+    /// The evaluation witness created by the prover in the interactive
+    /// protocol. Given a challenge (gamma_1 and gamma_2) and the claimed
+    /// evaluation of 2 sets of polynomials, the prover returns this witness,
+    /// proving the correctness of the evaluations. A verifier holding the
+    /// commitments for all polynomials in the 2 sets can then verify this
+    /// witness.
+    class evaluation_witness
+    {
+    public:
+        // Part of witness corresponding to evaluations of f_1s.
+        const libff::G1<ppT> W_1;
+        // Part of witness corresponding to evaluations of f_2s.
+        const libff::G1<ppT> W_2;
+
+        evaluation_witness(
+            const libff::G1<ppT> &W_1, const libff::G1<ppT> &W_2);
+    };
+
+    /// Evaluate the polynomials and return an evaluation object. f_s are
+    /// evaluated at z_1 and g_s are evaluated at z_2.
+    static evaluations evaluate_polynomials(
+        const std::vector<polynomial<Field>> &fs,
+        const std::vector<polynomial<Field>> &gs,
+        const Field &z_1,
+        const Field &z_2);
+
+    /// Generate a witness to prove the correctness of these evaluations.
+    /// Returns a `multi_point_evaluation_witness` object holding this data.
+    static evaluation_witness create_evaluation_witness(
+        const std::vector<polynomial<Field>> &fs,
+        const std::vector<polynomial<Field>> &gs,
+        const Field &z_1,
+        const Field &z_2,
+        const evaluations &evaluations,
```

Comment:
Same question as above w.r.t the need to pass the evaluation as argument here as opposed to running the evaluation as part of this function.

Reply:
I think it makes sense for the evaluations to be known to the verifier before choosing random gamma values.  This is how the scheme is given in Section 3 of [GWC19].  This was the main motivation for separating the evaluation in this scheme, and the kzg10 scheme was changed to fit with this.

Reply:
Hum sure, but this function is ran by the prover who's generating the witness to prove the "polynomialness" of the committed polynomial(s), at the previous steps, via the rational polynomial expression [f(x) - f(z)] / (x-z). So in essence, we could just require the prover to evaluate the polynomial(s) while generating the witness - instead of requiring the prover to pass an evaluation of the polynomials to this function.

Reply:
The proof in Section 3 of [GWC19] says explicitly that the evaluations `s_i` are available *before* the verifier chooses the random gammas.  It may be fine to change this, but I would need to go through a modified version of the proof to convince myself that the prover can see the gamma values *before* revealing the claimed evaluations.  For now, I think it's easiest to stick to the structure in the paper, in which the evaluations are revealed before gamma is chosen by the verifier.

Reply:
I am not suggesting to change anything w.r.t 

> the evaluations s_i are available before the verifier chooses the random gammas.

We are talking about the interface of the `create_evaluation_witness` function here, not the choice of gammas.

Reply:
These are different points, we can't justify that the evaluation is passed to the witness creation function because the evaluation needs to be passed to the verifier. I mean, the evaluated point needs to be sent to the verifier and we can just say: "we re-evaluate" phi(x) at point i in the witness generation function.
A good justification for not doing this would be: "well actually we can save a polynomial evaluation by just passing phi(i) as argument to the witness generation function". In which case: ok, makes sense, we're caching the evaluation for the message computation of the next iteration of the IP, so ok. This answers the question in https://github.com/clearmatics/libsnark/pull/42#discussion_r812035029 to the affirmative. Yes passing the evaluation to this function saves a polynomial evaluation.

But removing phi(i) from the witness generation function doesn't change the soundness.
In the IP, the alleged polynomial evaluation `s` is still sent to the Verifier (who's a RO in the non-interactive setting) and then the polynomial is re-evaluated in the witness generation.

Reply:
OK.  So you are suggesting we REcompute the evaluations inside the `create_evaluation_witness` function?  The prover should already have computed these values, so is this not redundant?  Especially if these functions are combined to create the non-interactive version.

Reply:
I'm not suggesting we do this, but was wondering why `phi(i)` (the evaluation) was passed as argument along with `phi` (the polynomial) and `i` (the evaluation point), as, at first look it seemed like a redundant piece of information to pass around (because the evaluation can be computed in the function body via `phi` and `i`). This was the nature of the question in: https://github.com/clearmatics/libsnark/pull/42#discussion_r812035029

Now I see that you passed it as argument to re-use the previously computed evaluation and thus to save an extra polynomial evaluation. So I understand the intent. That sounds like a good tradeoff to have a "redundant argument" in order to save an extra polynomial evaluation.

Reply:
> But removing phi(i) from the witness generation function arguments doesn't change the soundness.
In the IP, the alleged polynomial evaluation s is still sent to the Verifier (who's a RO in the non-interactive setting) and then the polynomial is re-evaluated in the witness generation by feeding the \phi and i arguments to the polynomial evaluation fucntion of lifqfft.

No, of course.  Whether the evaluations are reused or not does not affect the soundness.  I'm saying that: the evaluation must be computed *before* the witness generation, and so it makes sense to reuse them into this function.  

I see that they are not actually used in the current implementation for pure convenience, although there is a [TODO about this](https://github.com/clearmatics/libsnark/blob/poly-commit-multi-point/libsnark/polynomial_commitments/kzg10_batched.tcc#L169) in the code.  I'd rather have the interface in the known correct form, rather than change it later when we do the optimization to reuse the evaluations, so it's probably easiest if I just go back and do that now, and resolve the TODO.  This will also match the logic in the BDFG21 code in a later PR.

Reply:
> I'm saying that: the evaluation must be computed before the witness generation, and so it makes sense to reuse them into this function.

Yes, I see :)

Reply:
> so it's probably easiest if I just go back and do that now, and resolve the TODO. This will also match the logic in the BDFG21 code in a later PR.

Ok no problem, so I'll wait before merging

Reply:
OK, thanks @AntoineRondelet .
I just pushed the changes in the final commit of this branch.  Better to resolve that TODO up front (and it was actually more trivial than I expected).  Thanks.